core/trino-main/src/main/java/io/trino/sql/gen/PageFunctionCompiler.java

```diff
@@ -349,13 +349,11 @@ private MethodDefinition generateEvaluateMethod(
         BytecodeBlock body = method.getBody();
         Variable thisVariable = method.getThis();
 
-        declareBlockVariables(projection, page, scope, body);
-
         Variable wasNullVariable = scope.declareVariable("wasNull", body, constantFalse());
         RowExpressionCompiler compiler = new RowExpressionCompiler(
                 callSiteBinder,
                 cachedInstanceBinder,
-                fieldReferenceCompiler(callSiteBinder),
+                fieldReferenceCompilerProjection(callSiteBinder),
```

Comment:
does the same apply to filters?

Reply:
I had tried the same change for filters, but it somehow does not help there. The `Page#getBlock` method shows up in CPU profile in JFRs mostly from projection expressions rather than filters. I don't know the root cause for that though.